File: komik_scrape/manhuarml.py
import re
from bs4 import BeautifulSoup

# Wajib ada agar scraper.py bisa mendeteksi kecocokan URL
DOMAINS = ["manhuarmtl.com"]

def get_chapter_list(manga_url, fetch_func, headers):
    """
    Ekstrak daftar chapter dari halaman utama manga.
    Sesuai dengan fungsi fetchMangaChapters di JS.
    """
    try:
        # Menggunakan fungsi fetch_with_fallback bawaan dari scraper.py
        res = fetch_func(manga_url, headers)
        soup = BeautifulSoup(res.text, 'html.parser')
        
        chapters = []
        
        # Cari semua elemen li dengan class wp-manga-chapter
        for element in soup.select('li.wp-manga-chapter'):
            a_tag = element.select_one('a')
            if not a_tag:
                continue
                
            title = a_tag.text.strip()
            link = a_tag.get('href', '').strip()
            
            # Cari elemen i di dalam class chapter-release-date
            date_tag = element.select_one('.chapter-release-date i')
            date = date_tag.text.strip() if date_tag else ""
            
            if title and link:
                chapters.append({
                    "title": title,
                    "link": link,
                    "date": date
                })
                
        return chapters
    except Exception as e:
        print(f"[Error - ManhuaRMTL] Gagal mengambil daftar chapter: {e}")
        return []


def get_page_list(soup):
    """
    Ekstrak semua link gambar dari dalam chapter.
    Sesuai dengan fungsi fetchChapterImages di JS.
    """
    images = []
    
    # Cari img dengan class wp-manga-chapter-img di dalam class page-break
    for element in soup.select('.page-break img.wp-manga-chapter-img'):
        src = element.get('src')
        
        # Kadang website manga pakai atribut 'data-src' untuk lazy load
        if not src:
            src = element.get('data-src')
            
        if src:
            images.append(src.strip())
            
    return images


def get_chapter_name(soup):
    """
    Ekstrak dan pisahkan judul manga serta nomor chapter.
    Sesuai dengan fungsi fetchChapterTitleAndNumber di JS.
    """
    heading_tag = soup.select_one('h1#chapter-heading')
    raw_heading = heading_tag.text.strip() if heading_tag else "Unknown Title"
    
    manga_title = raw_heading
    chapter_number = ""
    
    # Regex untuk memisahkan teks sebelum dan sesudah kata Chapter/Chap/Bab
    # re.IGNORECASE memastikan tidak sensitif huruf besar/kecil
    regex = r'(.*?)\s*(?:Chapter|Chap|Bab)\s*(.*)'
    match = re.search(regex, raw_heading, re.IGNORECASE)
    
    if match:
        manga_title = match.group(1).strip()
        chapter_number = match.group(2).strip()
        
    return {
        "title": manga_title,
        "chapter_name": chapter_number
    }

File: komik_scrape/manhuarml.mjs
import * as cheerio from 'cheerio';

// Wajib diexport agar bisa dideteksi oleh scraper.mjs
export const DOMAINS = ["manhuarmtl.com"];

export async function getChapterList(url, fetchFunc) {
    const html = await fetchFunc(url);
    const $ = cheerio.load(html);
    const chapters = [];
    
    $('li.wp-manga-chapter').each((i, el) => {
        const title = $(el).find('a').text().trim();
        const link = $(el).find('a').attr('href');
        // Simpan sebagai 'url' agar sesuai dengan Python main.py
        if (title && link) chapters.push({ title: title, url: link }); 
    });
    return chapters;
}

export async function getPageList(html) {
    const $ = cheerio.load(html);
    const images = [];
    
    $('.page-break img.wp-manga-chapter-img').each((i, el) => {
        let src = $(el).attr('src') || $(el).attr('data-src');
        if (src) {
            // Wajib menggunakan index agar terurut di Python
            images.push({ index: i, url: src.trim() });
        }
    });
    return images;
}

export async function getChapterName(html) {
    const $ = cheerio.load(html);
    const rawHeading = $('h1#chapter-heading').text().trim();
    let mangaTitle = rawHeading;
    let chapterNumber = '';
    
    const match = rawHeading.match(/(.*?)\s*(?:Chapter|Chap|Bab)\s*(.*)/i);
    if (match) {
        mangaTitle = match[1].trim();
        chapterNumber = match[2].trim();
    }
    
    return { title: mangaTitle, chapter_name: chapterNumber };
}
